ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl.hs

```diff
@@ -180,233 +203,113 @@ forkSyncStateOnTipPointChange registry menv =
-            case headTxValidationRes of
-              -- Mempool capacity hasn't been reached (no remaining unvalidated
-              -- transactions were returned).
-              (vr, []) -> do
-                -- Continue validating the remaining transactions.
-                (vr', unvalidatedTxs) <- validateNewUntilMempoolFull remainingTxs vr
-                writeTVar mpEnvStateVar $ internalStateFromVR vr'
-                pure (vr', unvalidatedTxs)
-
-              -- The mempool capacity has been reached.
-              _ -> retry
-
-    validateNewUntilMempoolFull
-      :: [GenTx blk]
-      -- ^ The new transactions to validate.
-      -> ValidationResult blk
-      -- ^ The 'ValidationResult' from which to begin validating.
-      -> STM m (ValidationResult blk, [GenTx blk])
-      -- ^ The last 'ValidationResult' along with the remaining transactions
-      -- (those not yet validated due to the mempool capacity being reached).
-    validateNewUntilMempoolFull []        vr = pure (vr, [])
-    validateNewUntilMempoolFull (tx:txs') vr = do
-      -- Get the current mempool size
-      MempoolSize { msNumBytes = curSizeInBytes } <- getMempoolSize mpEnv
-
-      -- Determine what the mempool size would be if we were to commit the new
-      -- transactions we've validated thus far and also the next transaction
-      -- to validate, 'tx'.
-      -- If this value is greater than the 'MempoolCapacityBytes', then we
-      -- know not to continue validating at this time.
-      let newTxsBytes    = sum (txSize <$> vrNewValid vr) + txSize tx
-          newSizeInBytes = curSizeInBytes + newTxsBytes
-
-      -- The size of a mempool should never be greater than its capacity.
-      assert (curSizeInBytes <= mempoolCap) $
-        -- Here, we check whether we're at the mempool's capacity /before/
-        -- attempting to validate the next transaction.
-        if newSizeInBytes <= mempoolCap
-          then validateNewUntilMempoolFull txs' (extendVRNew mpEnvLedgerCfg tx vr)
-          else pure (vr, tx:txs') -- if we're at mempool capacity, we return the
-                                  -- last 'ValidationResult' as well as the
-                                  -- remaining transactions (those not yet
-                                  -- validated).
+    done acc toAdd = return (reverse acc, toAdd)
+
+    go acc []                     = done acc []
+    go acc toAdd@(firstTx:toAdd') = do
+      let firstTxSize = txSize firstTx
+      -- Note: we execute the continuation returned by 'atomically'
+      join $ atomically $ do
+        is <- readTVar mpEnvStateVar
+        let curSize = msNumBytes $ isMempoolSize is
+        if curSize + firstTxSize > capacity then
+          -- No space in the Mempool
+          return $ done acc toAdd
+        else do
+          let vr  = extendVRNew cfg firstTx $ validationResultFromIS is
+              is' = internalStateFromVR vr
+          unless (null (vrNewValid vr)) $
```

Comment:
With this new implementation, it looks like `vrNewValid` doesn't need to be a list since we're recursively validating the head transaction against a newly constructed `ValidationResult` that _always_ has a `vrNewValid` value of `[]`. Perhaps we should just define the `vrNewValid` field as a `Maybe`?

Reply:
Ah, good point! I thought that `revalidateTxsFor` still needed it, but I was wrong, it uses `vrValid` (through `extendVRPrevApplied`).

Although, thinking about it now, what should `extendVRNew` do when `vrNewValid` is a `Just`? Return an `error`, I suppose?

Reply:
Yeah, I think I would probably return an `error` or `assert` that it is `Nothing`.